lib/recursive-methods.rb

```diff
@@ -1,49 +1,99 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# source: https://www.bigocheatsheet.com/
+
+# Time complexity: O(n)
+# Space complexity: O(n)
+# Thank you to Chris-recursion video
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    raise ArgumentError if n < 0
+    return 1 if n == 0
+    return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(n)
+# Why can't I use .last?
+def reverse(s, first = 0, last = s.length-1)
+    return s if first > last
+
+    holder = s[first]
+    s[first] = s[last]
+    s[last] = holder
+
+    return reverse(s, first +1, last -1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
+# source: https://medium.com/@txea/reverse-a-string-with-recursion-a3f8505ffb73
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+    return s if s.length <= 1
+    gnirts = reverse_inplace(s[1..-1])
+    gnirts += s[0]
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+    return 0 if n == 0
+    return 2 + bunny(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
+# Talk to tutor
 def nested(s)
```

Comment:
Since you are making a new recursive call with each method, this is O(n) space complexity ( call stack), and `s.slice!(0)` is O(n) alone.  So your time complexity is O(n^2).